pkg/envtest/komega/matcher.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package komega
+
+import (
+	"context"
+	"time"
+
+	"github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Matcher has Gomega Matchers that use the controller-runtime client.
+type Matcher struct {
+	Client       client.Client
+	ctx          context.Context
+	extras       []interface{}
+	timeout      time.Duration
+	pollInterval time.Duration
+}
+
+// WithContext sets the context to be used for the underlying client
+// during assertions.
+func (m *Matcher) WithContext(ctx context.Context) Komega {
+	m.ctx = ctx
+	return m
+}
+
+// context returns the matcher context if one has been set.
+// Else it returns the context.TODO().
+func (m *Matcher) context() context.Context {
+	if m.ctx == nil {
+		return context.TODO()
+	}
+	return m.ctx
+}
+
+// WithExtras sets extra arguments for sync assertions.
+// Any extras passed will be expected to be nil during assertion.
+func (m *Matcher) WithExtras(extras ...interface{}) KomegaSync {
+	m.extras = extras
+	return m
+}
+
+// WithTimeout sets the timeout for any async assertions.
+func (m *Matcher) WithTimeout(timeout time.Duration) KomegaAsync {
+	m.timeout = timeout
+	return m
+}
+
+// WithPollInterval sets the poll interval for any async assertions.
+// Note: This will only work if an explicit timeout has been set with WithTimeout.
+func (m *Matcher) WithPollInterval(pollInterval time.Duration) KomegaAsync {
+	m.pollInterval = pollInterval
+	return m
+}
+
+// intervals constructs the intervals for async assertions.
+// If no timeout is set, the list will be empty.
+func (m *Matcher) intervals() []interface{} {
+	if m.timeout == 0 {
+		return []interface{}{}
+	}
+	out := []interface{}{m.timeout}
+	if m.pollInterval != 0 {
+		out = append(out, m.pollInterval)
+	}
+	return out
+}
+
+// Create creates the object on the API server.
+func (m *Matcher) Create(obj client.Object, opts ...client.CreateOption) gomega.GomegaAssertion {
+	err := m.Client.Create(m.context(), obj, opts...)
+	return gomega.Expect(err, m.extras...)
```

Comment:
Currently, these will only work with the default package level gomega assertions, should make the option of passing in a custom `gomega.WithT` as well to this can be used in normal go testing tests